core/primitives/src/block.rs

```diff
@@ -340,15 +342,31 @@ impl Block {
         );
 
         let new_total_supply = prev.total_supply() + minted_amount.unwrap_or(0) - balance_burnt;
-        let now = clock.now_utc().unix_timestamp_nanos() as u64;
-        #[cfg(feature = "sandbox")]
-        let now = now + sandbox_delta_time.unwrap().whole_nanoseconds() as u64;
-        #[cfg(not(feature = "sandbox"))]
-        debug_assert!(sandbox_delta_time.is_none());
-        let time = if now <= prev.raw_timestamp() { prev.raw_timestamp() + 1 } else { now };
-
-        let (vrf_value, vrf_proof) = signer.compute_vrf_with_proof(prev.random_value().as_ref());
-        let random_value = hash(vrf_value.0.as_ref());
+
+        // Use the optimistic block data if available, otherwise compute it.
+        let (time, vrf_value, vrf_proof, random_value) = optimistic_block
+            .as_ref()
+            .map(|ob| {
+                (
+                    ob.inner.block_timestamp,
+                    ob.inner.vrf_value,
+                    ob.inner.vrf_proof,
+                    ob.inner.random_value,
+                )
+            })
+            .unwrap_or_else(|| {
+                let now = clock.now_utc().unix_timestamp_nanos() as u64;
+                #[cfg(feature = "sandbox")]
+                let now = now + sandbox_delta_time.unwrap().whole_nanoseconds() as u64;
+                #[cfg(not(feature = "sandbox"))]
+                debug_assert!(sandbox_delta_time.is_none());
+                let time = if now <= prev.raw_timestamp() { prev.raw_timestamp() + 1 } else { now };
+
+                let (vrf_value, vrf_proof) =
+                    signer.compute_vrf_with_proof(prev.random_value().as_ref());
+                let random_value = hash(vrf_value.0.as_ref());
+                (time, vrf_value, vrf_proof, random_value)
```

Comment:
Please make a common function to avoid duplicating of `OptimisticBlock::produce` logic.